### PR TITLE
oauthutil: support IPv6 auth server listen bind address

### PR DESCRIPTION
**Pull request by tonymet**
#### What is the purpose of this change?
Support IPv6 Bind Address for Oauth authentication server.   This allows for e2e IPv6 connections. 
* Works for machines with only IPv6 interface (e.g. EC2 without public IPv4 address)
* Supports IPv6 SSH tunnel with `ssh -6 -L53682:localhost:53682`

Added new config param `config_bind_address` 

e.g. 
```
[drive-test]
type = drive
config_bind_address = [::1]:53682
```

#### Other Approaches to Consider
* Re-use the existing `--bind flag` to set the local address accordingly.  
   * This was a bit kludgy because --bind takes a hostname and would require clumsy parsing of the name to identify IPv6
 * Bind to Both interfaces? 
    * this seemed over complicated and would require listening and acting on two http servers

#### Part of IPv6 E2E support for Rclone:  
* ✅--bind to IPv6 for outbound requests
* [THIS PR] ➡️oauth using IPv6 (::1)
* ✅Connect to S3 endpoint on IPv6 (see PR  #7482) 
* ✅ Connect to Google Drive API endpoint on IPv6 (once authorized )
* ❓Connect to other backend endpoints on IPv6

#### Was the change discussed in an issue or in the forum before?
https://forum.rclone.org/t/ipv6-guide-review-with-google-drive-ec2-s3/43218

#### Testing
1. Run `go run . config` and configure "drive" backend with

``` 
$ go run . config 
2024/01/01 18:09:48 NOTICE: Make sure your Redirect URL is set to "http://[::1]:53682/" in your custom config.
2024/01/01 18:09:48 NOTICE: If your browser doesn't open automatically go to the following link: http://[::1]:53682/auth?state=Wx_SmigfWPVEfstjTxvRig
2024/01/01 18:09:48 NOTICE: Log in and authorize rclone for access
2024/01/01 18:09:48 NOTICE: Waiting for code...

```
Hit CTRL-Z to background the process

Run `netstat` (below) to confirm `tcp6` 

You should see `tcp6` and `"[::]:*"` in `netstat` when binding with ipv6 address
```
$  sudo netstat -tulp|grep rclone 
tcp6       0      0 localhost:53682         [::]:*                  LISTEN      21555/rclone   
```

#### Checklist

- [ ✅] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
-   [ ] I have added tests for all changes in this PR if appropriate.
- [ ✅] I have added documentation for the changes if appropriate.
- [ ✅] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Review by ncw — COMMENTED** [review]
Here is an idea...

What do you think of using `localhost` but interpreting it ourselves as either 127.0.0.1 or ::1 whichever binds first. We can then use that in the URL.

This would save having to configure the bind address explicitly, and having to rely on `localhost` resolving to something sensible which isn't a given.

Thinking more I think that people have asked for the bind address of the oauth server to be configurable - you'll probably find an issue about that!

File: lib/oauthutil/oauthutil.go [hunk 1]
```diff
@@ -102,6 +100,21 @@ var SharedOptions = []fs.Option{{
 	Name:     config.ConfigTokenURL,
 	Help:     "Token server url.\n\nLeave blank to use the provider defaults.",
 	Advanced: true,
+}, {
+	Name:    config.ConfigBindAddress,
```
[the remark this answers is not in the document]
Comment (tonymet) [reply lib/oauthutil/oauthutil.go:104]: I agree with you it's clumsy (see "considerations").  I'd also like to make this more automatic.  Some ideas ... 

1. depend on "--bind" and listen on the same interface
2. always use "http://localhost..." and let the NS lookup handle the interface binding. 
3. Listen on both (dual read or proxy).

File: lib/oauthutil/oauthutil.go [hunk 2]
```diff
@@ -28,6 +28,10 @@ import (
 var (
 	// templateString is the template used in the authorization webserver
 	templateString string
+	// bindAddress is binding for local webserver when active
+	bindAddress string = "localhost:" + bindPort
```
[the remark this answers is not in the document]
Comment (tonymet) [reply lib/oauthutil/oauthutil.go:32]: it's a fair concern. I used it to allow for automatic configuration on IPv6 only machines (nslookup "localhost" will return `::1` )

**Comment by tonymet** [comment]
> Here is an idea...
> 
> What do you think of using `localhost` but interpreting it ourselves as either 127.0.0.1 or ::1 whichever binds first. We can then use that in the URL.

Do you mean roughly ....
```
if (net.Listen(::1))
// listen on ipv6
 else net.listen(127.0.0.1)
// listen on ipv4

```

or (current PR) 
```
BindAddr = nslookup("localhost")
net.listen(BindAddr) 
```

**Comment by tonymet** [comment]
> Thinking more I think that people have asked for the bind address of the oauth server to be configurable - you'll probably find an issue about that!

one benefit is this  Allows for oauth redirect to remote machine ( e.g. rclone running on cloud instance, web browser running on laptop)

File: lib/oauthutil/oauthutil.go [hunk 3]
```diff
@@ -849,10 +846,14 @@ func (s *authServer) Init() error {
 	mux.HandleFunc("/", s.handleAuth)
 
 	var err error
-	s.listener, err = net.Listen("tcp", s.bindAddress)
-	if err != nil {
-		return err
+	if s.listener, err = net.Listen("tcp", "[::1]:"+bindPort); err != nil {
```
Comment (tonymet) [inline lib/oauthutil/oauthutil.go:849]: @ncw do you like this approach better? I'll still need to clean up some of the duplicate state like authUrl, redirectUrl, oauthConfig etc.

File: lib/oauthutil/oauthutil.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tonymet) [reply lib/oauthutil/oauthutil.go:32]: i'll update this to use `s.listener.Addr().String()` when i clean it up

File: lib/oauthutil/oauthutil.go (hunk 3 above)
Comment (ncw) [reply lib/oauthutil/oauthutil.go:849]: Yes.

Does `[::1]` include `127.0.0.1`? If not then we need to bind to both if possible which is a little awkward!

Comment (tonymet) [reply lib/oauthutil/oauthutil.go:849]: I'm still researching options to listen to both tcp4 and tcp6.   From net.Listen docs we can either (a) listen on a specific address or (b) listen on all multicast/unicast (not safe)

Here are the options I'm aware of to listen to ::1 & 127.0.0.1
 
1. Implement a MultiListener interface in GO (about 50 lines -- a common approach)
2. spawn (2) net.LIsten & http.Serve objects with the identical handler routine
3. TBD -- find the appropriate SetSockOpt / SockAddr struct config to listen on ::1 & 127.0.0.1 but not external interfaces

 I'm continuing research on (3) by reviewing net.go implementation and the C structs man pages.

**Pull request closed without merging**